addon/-private/system/relationships/relationships-payloads.js

```diff
@@ -0,0 +1,417 @@
+    this._store = store;
+    this._lhsModelName = modelName;
+    this._lhsRelationshipName = relationshipName;
+    this._lhsRelationshipMeta = relationshipMeta;
+    this._rhsModelName = inverseModelName;
+    this._rhsRelationshipName = inverseRelationshipName;
+    this._rhsRelationshipMeta = inverseRelationshipMeta;
+
+    this._lhsPayloads = {};
+    if (modelName !== inverseModelName || relationshipName !== inverseRelationshipName) {
+      this._rhsPayloads = {};
+      this._isReflexive = false;
+    } else {
+      // Edge case when we have a reflexive relationship to itself
+      //  eg user hasMany friends inverse friends
+      this._rhsPayloads = this._lhsPayloads;
+      this._isReflexive = true;
+    }
+
+    // either canoical on push or pending & flush
+    this._pendingPayloads = [];
+  }
+
+  get(modelName, id, relationshipName) {
+    this._flushPending();
+
+    if (modelName === this._lhsModelName && relationshipName === this._lhsRelationshipName) {
+      return this._lhsPayloads[id];
+    } else {
+      return this._rhsPayloads[id];
+    }
+  }
+
+  push(modelName, id, relationshipName, relationshipData) {
+    this._pendingPayloads.push([modelName, id, relationshipName, relationshipData]);
+  }
+
+  unload(modelName, id, relationshipName) {
+    this._flushPending();
+
+    if (modelName === this._lhsModelName && relationshipName === this._lhsRelationshipName) {
+      this._unload(id, this._lhsPayloads, this._rhsPayloads);
+    } else {
+      this._unload(id, this._rhsPayloads, this._lhsPayloads);
+    }
+  }
+
+  _unload(id, payloads, inversePayloads) {
+    if (!this._inverseLoaded(payloads[id])) {
+      this._removeInverse(id, payloads[id], inversePayloads);
+      delete payloads[id];
+    }
+  }
+
+  _flushPending() {
+    if (this._pendingPayloads.length === 0) { return; }
+
+    let work = this._pendingPayloads.splice(0, this._pendingPayloads.length);
+    for (let i=0; i<work.length; ++i) {
+      let modelName = work[i][0];
+      let id = work[i][1];
+      let relationshipName = work[i][2];
+      let relationshipData = work[i][3];
+
+      // TODO: maybe delay this slightly?
+      let inverseRelationshipData = {
+        data: {
+          id: id,
+          type: modelName
+        }
+      }
+
+      if (modelName === this._lhsModelName && relationshipName === this._lhsRelationshipName) {
+        this._removeInverse(id, this._lhsPayloads[id], this._rhsPayloads);
+        this._lhsPayloads[id] = relationshipData;
+        this._populateInverse(relationshipData, inverseRelationshipData, this._rhsPayloads, this._rhsRelationshipIsMany);
+      } else {
+        this._removeInverse(id, this._rhsPayloads[id], this._lhsPayloads);
+        this._rhsPayloads[id] = relationshipData;
+        this._populateInverse(relationshipData, inverseRelationshipData, this._lhsPayloads, this._lhsRelationshipIsMany);
+      }
+    }
+  }
+
+  _inverseLoaded(payload) {
```

Comment:
Seems bad that the invariant from unloading where graphs are either unloaded together or not at all, is changed here